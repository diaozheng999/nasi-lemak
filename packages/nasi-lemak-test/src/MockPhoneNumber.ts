/**
 * MockPhoneNumber.ts
 * @author Diao Zheng
 * @file Phone number generation
 */

// @barrel export all

import { Option } from "nasi-lemak";

export interface IMockPhoneNumberSpecifications {
  prefixList: string[];
  numberOfDigits: number;
  format?: (unformatted: string) => string;
}

const a = 1073740463;
const b = 1073737877;
const c = 1073739059;

const MAX_DIGIT_PER_ITERATION = 9;

const digitMask = [
  /* 0 */ 1,
  /* 1 */ 10,
  /* 2 */ 100,
  /* 3 */ 1000,
  /* 4 */ 10000,
  /* 5 */ 100000,
  /* 6 */ 1000000,
  /* 7 */ 10000000,
  /* 8 */ 100000000,
  /* 9 */ 1000000000,
];

function generateDigit(
  seed: number,
  digits: number,
  offset: number = 0,
): string {
  if (digits <= MAX_DIGIT_PER_ITERATION) {
    const mask = digitMask[digits];
    return Math.abs(
      ((a + (seed * b) + (offset * c)) % mask) + mask).toString(10).slice(1);
  } else {
    return generateDigit(seed, digits - MAX_DIGIT_PER_ITERATION, offset + 1);
  }
}

/**
 * Generate a "believable" phone number based on an auto-increasing sequence.
 * The sequence goes as follows:
 * - generate(0) = 83740463
 * - generate(1) = 97478340
 * - generate(2) = 81216217
 * - generate(3) = 94954094
 * - generate(4) = 88691971
 * - generate(5) = 92429848
 * - generate(6) = 86167725
 * - generate(7) = 99905602
 * - generate(8) = 83643479
 * - generate(9) = 97381356
 *
 * This is similar to the "pseudo-encrypt" algorithm used to store unique
 * primary keys, and is guaranteed to be the same sequence.
 *
 * The first number (8/9) depends on the parity of `k`. The rest is a pseudo-
 * random sequence.
 *
 * @param seed a auto-increasing number
 * @returns a 8-character string that's a believable phone number.
 */
export function generate(
  seed: number,
  spec: IMockPhoneNumberSpecifications = SG,
) {
  const prefix = spec.prefixList[seed % spec.prefixList.length];
  const phoneNumber = prefix + generateDigit(seed, spec.numberOfDigits);
  return Option.callIf(spec.format, phoneNumber);
}

export function format(strings: TemplateStringsArray, ...counts: number[]) {
  return (phoneNumber: string) => {
    let response = strings[0];
    let start = 0;
    for (let i = 0; i < counts.length; ++i) {
      const end = start + counts[i];
      response += phoneNumber.slice(start, end) + strings[i + 1];
      start = end;
    }
    return response;
  };
}

export function unformatted(
  spec: IMockPhoneNumberSpecifications,
): IMockPhoneNumberSpecifications {
  // tslint:disable-next-line: no-shadowed-variable
  const { format, ...others } = spec;
  return others;
}

export const SG: IMockPhoneNumberSpecifications = {
  format: format`${4} ${4}`,
  numberOfDigits: 7,
  prefixList: ["8", "9"],
};
