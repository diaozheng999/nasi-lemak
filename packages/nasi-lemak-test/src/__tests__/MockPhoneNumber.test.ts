/**
 * MockPhoneNumber.test.ts
 * @author Diao Zheng
 * @file snapshot to ensure that phone number generated is identical.
 */

import { generate, SG, unformatted } from "../MockPhoneNumber";

test("phone number generator", () => {

  const numbers = [];

  const spec = unformatted(SG);

  for (let i = 0; i < 100; ++i) {
    expect(generate(i, spec)).toBe(generate(i, spec));
    expect(generate(i, spec)).toHaveLength(8);
    numbers.push(generate(i, spec));
  }
  expect(numbers).toMatchSnapshot();
});
