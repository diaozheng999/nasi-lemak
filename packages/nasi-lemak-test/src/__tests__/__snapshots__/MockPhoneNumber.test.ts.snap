// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`phone number generator 1`] = `
Array [
  "83740463",
  "97478340",
  "81216217",
  "94954094",
  "88691971",
  "92429848",
  "86167725",
  "99905602",
  "83643479",
  "97381356",
  "81119233",
  "94857110",
  "88594987",
  "92332864",
  "86070741",
  "99808618",
  "83546495",
  "97284372",
  "81022249",
  "94760126",
  "88498003",
  "92235880",
  "85973757",
  "99711634",
  "83449511",
  "97187388",
  "80925265",
  "94663142",
  "88401019",
  "92138896",
  "85876773",
  "99614650",
  "83352527",
  "97090404",
  "80828281",
  "94566158",
  "88304035",
  "92041912",
  "85779789",
  "99517666",
  "83255543",
  "96993420",
  "80731297",
  "94469174",
  "88207051",
  "91944928",
  "85682805",
  "99420682",
  "83158559",
  "96896436",
  "80634313",
  "94372190",
  "88110067",
  "91847944",
  "85585821",
  "99323698",
  "83061575",
  "96799452",
  "80537329",
  "94275206",
  "88013083",
  "91750960",
  "85488837",
  "99226714",
  "82964591",
  "96702468",
  "80440345",
  "94178222",
  "87916099",
  "91653976",
  "85391853",
  "99129730",
  "82867607",
  "96605484",
  "80343361",
  "94081238",
  "87819115",
  "91556992",
  "85294869",
  "99032746",
  "82770623",
  "96508500",
  "80246377",
  "93984254",
  "87722131",
  "91460008",
  "85197885",
  "98935762",
  "82673639",
  "96411516",
  "80149393",
  "93887270",
  "87625147",
  "91363024",
  "85100901",
  "98838778",
  "82576655",
  "96314532",
  "80052409",
  "93790286",
]
`;
